### DISCUSSION TITLE
Enchancement: Automatically cleanup unused docker images

### DISCUSSION THREAD

**Discussion by Hedius** [IceWhaleTech/CasaOS#251] [Ideas]
Discussion on discord: https://discord.com/channels/884667213326463016/981983949402292274 

It seems like casaOS does not delete unused docker images?

At least a user had the problem that his disk was full.

docker image prune was able to free up 5GB of unused docker images.

Idea: maybe automatically remove unused docker images?

A cronjob once per night for example would be a quick and dirty fix for that: docker image prune --all

**Comment by Protektor-Desura** [comment]
If you use something like watchtower to update your dockers then you can have it automatically remove old unused images.

```version: '3.3'
services:
  watchtower:
    command: --cleanup --rolling-restart --include-stopped
    container_name: watchtower
    environment:
      - TZ=America/Chicago
    hostname: watchtower
    image: containrrr/watchtower:latest
    networks:
      - archon
    restart: unless-stopped
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:rw"


networks:
  archon:
    external: true```

**Comment by Hedius** [comment]
That wasn't my point.

My point is that by default casaos does not delete images and it should do that. Without a 3rd party service in the background.

Especially because people on the discord suddenly have full disks without knowing why.

**Reply by gjchester** [reply]
Isn't that a design choice by Docker too?   Docker by default will not remove unused images, and there may be cases where you will not want to auto clean everything (say low bandwidth sites) .

Maybe a better option is a button/ action to trigger the full Docker purge, (ie a systme prune)  on user request rather than an overnight repeating process.

**Reply by Hedius** [reply]
> Isn't that a design choice by Docker too? Docker by default will not remove unused images, and there may be cases where you will not want to auto clean everything (say low bandwidth sites) .
> 
> Maybe a better option is a button/ action to trigger the full Docker purge, (ie a systme prune) on user request rather than an overnight repeating process.

If a user deletes an app, then a user expects that the app is fully deleted. And not that the disk gets full slowly /var/lib/docker without them knowing why.
You cannot expect from users to understand how containers work. (See the discord).

**Reply by Protektor-Desura** [reply]
> Isn't that a design choice by Docker too? Docker by default will not remove unused images, and there may be cases where you will not want to auto clean everything (say low bandwidth sites) .

Yes. There are times when an update is a problem so you end up wanting to roll back to a previous version that is not buggy.

**Reply by Protektor-Desura** [reply]
> > Isn't that a design choice by Docker too? Docker by default will not remove unused images, and there may be cases where you will not want to auto clean everything (say low bandwidth sites) .
> > Maybe a better option is a button/ action to trigger the full Docker purge, (ie a systme prune) on user request rather than an overnight repeating process.
> 
> If a user deletes an app, then a user expects that the app is fully deleted. And not that the disk gets full slowly /var/lib/docker without them knowing why. You cannot expect from users to understand how containers work. (See the discord).

I would think the web app should ask if they are removing an app if they want to remove the config directory and the image.

The reason I say ask about both is there are times where basically they may want to shut down an app/service at the moment because of various reasons but don't want all their configurations blow away because they will run it again in the future. I can see also where low bandwidth places might not want to have to download the container image again as well.

Maybe prompt if they want to remove everything and if they say no (not cancel the removal) then ask the next two questions of what parts to remove (config & container image).

**Reply by Hedius** [reply]
Containers and images should never contain any configuration. If you want to preserve the configuration people should just use volumes. Promoting anything else is totally wrong.

**Reply by Hedius** [reply]
* or use environment variables.

**Reply by Hedius** [reply]
I don't get your argument honestly. If you want to preserve it just stop the container. Then the image is also still used.

Regarding low bandwidth. A config switch can be used for that.

Asking users every time they delete the container is decreasing the user experience in my opinion.

**Reply by Protektor-Desura** [reply]
> Containers and images should never contain any configuration. If you want to preserve the configuration people should just use volumes. Promoting anything else is totally wrong.

No you don't save your configuration and settings for the server/app in a volume. You should be saving them in a mounted directory to easier back them up and edit the files easier if needed.

**Reply by Protektor-Desura** [reply]
There are three parts to the docker, the config directory, the image and then running docker part. If you are using yml files then there is that part as well.

**Reply by Hedius** [reply]
Docker compose has nothing to do with this topic.... and please don't try to explain containers to me.

**Reply by Protektor-Desura** [reply]
Then don't say you don't get my argument and then get nasty when I explain my reasoning.

**Reply by gjchester** [reply]
> If a user deletes an app, then a user expects that the app is fully deleted. And not that the disk gets full slowly /var/lib/docker without them knowing why.
You cannot expect from users to understand how containers work. (See the discord).

I can totally understand what you mean, but while new users may expect docker containers to be removed after the app is deleted, thats not expected Docker behaviour and experienced docker users may be confused by such actions.

Maybe the solution lies in the App Delete function of CasaOS runs a docker purge as part of its process?

**Comment by AuthorShin** [comment]
As I said [here](https://github.com/IceWhaleTech/CasaOS/discussions/190) just add a button to switch between advanced and basic mode and then add options like "Automatically clean up unused docker images" under basic or advanced mode so users can choose from.
What casaOS needs, is to give the users more choice power.
